<a href="https://colab.research.google.com/github/kuroshkarimi/Machine-Learning-and-Data-Analysis/blob/main/Pipeline_regression2/Power%20Plant_different%20regressions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Combined Cycle Power Plant
# Task Statement
You are provided with a dataset from the UCI Machine Learning Repository that contains measurements from a power plant.
the data is deposited in my github and link to its raw version is:

("https://github.com/kuroshkarimi/Machine-Learning-and-Data-Analysis/raw/refs/heads/main/Pipeline_regression2/Combined%20Cycle%20Power%20Plant.csv")


The objective is to predict the energy output (PE) of the power plant based on the following technical features:

AT: Ambient Temperature

V: Exhaust Vacuum

AP: Ambient Pressure

RH: Relative Humidity

Using this dataset build regression models to predict the energy output (PE) using the Pipeline class and various regression estimators.

# You should:

Preprocess the data as needed (e.g., scaling, splitting).

Use different estimators (e.g., LinearRegression, RandomForestRegressor, etc.).

Evaluate and compare model performance using appropriate regression metrics (e.g., RMSE, R²).





In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
url = "https://github.com/kuroshkarimi/Machine-Learning-and-Data-Analysis/raw/refs/heads/main/Pipeline_regression2/Combined%20Cycle%20Power%20Plant.csv"
data = pd.read_csv(url)
data.head()

,AT,V,AP,RH,PE
0,14.96,41.76,1024.07,73.17,463.26
1,25.18,62.96,1020.04,59.08,444.37
2,5.11,39.40,1012.16,92.14,488.56
3,20.86,57.32,1010.24,76.64,446.48
4,10.82,37.50,1009.23,96.62,473.90


In [3]:
# Consideration of the table characteristics

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9568 entries, 0 to 9567
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   AT      9568 non-null   float64
 1   V       9568 non-null   float64
 2   AP      9568 non-null   float64
 3   RH      9568 non-null   float64
 4   PE      9568 non-null   float64
dtypes: float64(5)
memory usage: 373.9 KB


In [4]:
# Inspection of the statistical properties of the columns

data.describe()

,AT,V,AP,RH,PE
count,9568.000000,9568.000000,9568.000000,9568.000000,9568.000000
mean,19.651231,54.305804,1013.259078,73.308978,454.365009
std,7.452473,12.707893,5.938784,14.600269,17.066995
min,1.810000,25.360000,992.890000,25.560000,420.260000
25%,13.510000,41.740000,1009.100000,63.327500,439.750000
50%,20.345000,52.080000,1012.940000,74.975000,451.550000
75%,25.720000,66.540000,1017.260000,84.830000,468.430000
max,37.110000,81.560000,1033.300000,100.160000,495.760000


In [8]:
# Separation of features from the target column and converting them to numpy arrays

X = data.iloc[:, :-1].values
y = data.iloc[:, -1].values

In [9]:
# Splitting the dataset into train and test segments
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [26]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.metrics import r2_score, root_mean_squared_error, mean_absolute_error


In [21]:
# 1. Linear Regression Model

pip_lin_reg = Pipeline(steps = [('scaler', StandardScaler()), ('model', LinearRegression())])
pip_lin_reg.fit(X_train, y_train)
y_lin_reg = pip_lin_reg.predict(X_test)
r2_lin_reg = r2_score(y_test, y_lin_reg)
rmse_lin_reg = root_mean_squared_error(y_test, y_lin_reg)
mae_lin_regr = mean_absolute_error(y_test, y_lin_reg)

print('\n r2_lr: ', r2_lin_regression)
print('RMSE_lr: ', rmse_lin_reg)
print('MAE_lr: ', mae_lin_regr)


 r2_lr:  0.9325315554761302
RMSE_lr:  4.442262858442491
MAE_lr:  3.566564655203821


In [22]:
# 2. Random Forest regression Model
pip_rf = Pipeline(steps = [('scaler', StandardScaler()), ('model', RandomForestRegressor(n_estimators = 100))])
pip_rf.fit(X_train, y_train)
y_rf = pip_rf.predict(X_test)
r2_rf = r2_score(y_test, y_rf)
rmse_rf = root_mean_squared_error(y_test, y_rf)
mae_rf = mean_absolute_error(y_test, y_rf)

print('\n r2_rf: ', r2_rf)
print('RMSE_rf: ', rmse_rf)
print('MAE_rf: ', mae_rf)



 r2_rf:  0.9646656739733692
RMSE_rf:  3.2147877905802615
MAE_rf:  2.3488653082549673


In [25]:
# 3. Support vector regression model
pip_sv = Pipeline(steps = [('scaler', StandardScaler()), ('model', SVR(kernel = 'rbf'))])
pip_sv.fit(X_train, y_train)
y_sv = pip_sv.predict(X_test)
r2_sv = r2_score(y_test, y_sv)
rmse_sv = root_mean_squared_error(y_test, y_sv)
mae_sv = mean_absolute_error(y_test, y_sv)

print('\n r2_sv: ', r2_sv)
print('RMSE_sv: ', rmse_sv)
print('MAE_sv: ', mae_sv)


 r2_sv:  0.9430195908033736
RMSE_sv:  4.082410630002811
MAE_sv:  3.1433150718242002


In [28]:
# 4. Gradient Boosting Regression model
pip_gb = Pipeline(steps = [('scaler', StandardScaler()), ('model', GradientBoostingRegressor())])
pip_gb.fit(X_train, y_train)
y_gb = pip_gb.predict(X_test)
r2_gb = r2_score(y_test, y_gb)
rmse_gb = root_mean_squared_error(y_test, y_gb)
mae_gb = mean_absolute_error(y_test, y_gb)

print('\n r2_gb: ', r2_gb)
print('RMSE_gb: ', rmse_gb)
print('MAE_gb: ', mae_gb)


 r2_gb:  0.9524627717123879
RMSE_gb:  3.7288149602286724
MAE_gb:  2.8902254135976473


## Importing the libraries

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Importing the dataset

In [ ]:
dataset = pd.read_csv(r"C:\workFile\Machine Learning\S01 - Machine-Learning-A-Z-Codes-Datasets\Machine-Learning-A-Z-Codes-Datasets\Part 2 - Regression\project regression 5\Combined Cycle Power Plant.csv")

dataset.head()

,AT,V,AP,RH,PE
0,14.96,41.76,1024.07,73.17,463.26
1,25.18,62.96,1020.04,59.08,444.37
2,5.11,39.40,1012.16,92.14,488.56
3,20.86,57.32,1010.24,76.64,446.48
4,10.82,37.50,1009.23,96.62,473.90


In [ ]:
# __ Evaluation of the data __
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9568 entries, 0 to 9567
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   AT      9568 non-null   float64
 1   V       9568 non-null   float64
 2   AP      9568 non-null   float64
 3   RH      9568 non-null   float64
 4   PE      9568 non-null   float64
dtypes: float64(5)
memory usage: 373.9 KB


In [ ]:
# __ Brief statistical analysis of the data __
dataset.describe()

,AT,V,AP,RH,PE
count,9568.000000,9568.000000,9568.000000,9568.000000,9568.000000
mean,19.651231,54.305804,1013.259078,73.308978,454.365009
std,7.452473,12.707893,5.938784,14.600269,17.066995
min,1.810000,25.360000,992.890000,25.560000,420.260000
25%,13.510000,41.740000,1009.100000,63.327500,439.750000
50%,20.345000,52.080000,1012.940000,74.975000,451.550000
75%,25.720000,66.540000,1017.260000,84.830000,468.430000
max,37.110000,81.560000,1033.300000,100.160000,495.760000


In [ ]:
# __ Separation of the predictors and target columns __

X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

In [ ]:
# we can have look at the features and the target columns
print(f'The predictors: \n {X}\n')
print(f'The target: {y}')

The predictors: 
 [[  14.96   41.76 1024.07   73.17]
 [  25.18   62.96 1020.04   59.08]
 [   5.11   39.4  1012.16   92.14]
 ...
 [  31.32   74.33 1012.92   36.48]
 [  24.48   69.45 1013.86   62.39]
 [  21.6    62.52 1017.23   67.87]]

The target: [463.26 444.37 488.56 ... 429.57 435.74 453.28]


In [ ]:
# __ Splitting the data into train and test sections

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

In [ ]:
# __ fitting the data with multiple linear regressive model and its r2 parameter
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, root_mean_squared_error, mean_absolute_error

# creation of the pipeline object with standard scaler normalization and multiple linear regression estimator
pipe_lin = Pipeline(steps=[('scaler', StandardScaler()), ('model', LinearRegression())])
pipe_lin.fit(X_train, y_train)
y_lin = pipe_lin.predict(X_test)
r2_lin = pipe_lin.score(X_test, y_test)
rmse_lin = -root_mean_squared_error(y_test, y_lin)
mae_lin = -mean_absolute_error(y_test, y_lin)
print(f'Multiple Linear Regression R-Squared = {r2_lin:.3f}')
print(f'negative Multiple Linear Regression RMSE = {rmse_lin:.3f}')
print(f'negative Multiple Linear Regression MAE = {mae_lin:.3f}')

Multiple Linear Regression R-Squared = 0.932
negative Multiple Linear Regression RMSE = -4.453
negative Multiple Linear Regression MAE = -3.561


In [ ]:
# __ fitting the data with a degree 2 polynomial linear regressive model and its r2 parameter

from sklearn.preprocessing import PolynomialFeatures

# creation of the pipeline object with standard scaler normalization, degree 2 polynomial feature transformation and multiple linear regression estimator
pipe_poly = Pipeline(steps = [('scaler', StandardScaler()), ('poly', PolynomialFeatures(degree = 2)), ('model', LinearRegression())])
pipe_poly.fit(X_train, y_train)
y_poly = pipe_poly.predict(X_test)
r2_poly= pipe_poly.score(X_test, y_test)
rmse_poly = -root_mean_squared_error(y_test, y_poly)
mae_poly = -mean_absolute_error(y_test, y_poly)
print(f'Polynomial Linear Regression R-Squared = {r2_poly:.3f}')
print(f'negative Polynomial  Linear Regression RMSE = {rmse_poly:.3f}')
print(f'negative Polynomial  Linear Regression MAE = {mae_poly:.3f}')

Polynomial Linear Regression R-Squared = 0.942
negative Polynomial  Linear Regression RMSE = -4.121
negative Polynomial  Linear Regression MAE = -3.285


In [ ]:
# __ fitting the data with a support vector regressive model and its r2 parameter

from sklearn.svm import SVR

# creation of the pipeline object with standard scaler normalization, and SVR regressive estimator with a linear kernel
pipe_svr = Pipeline(steps = [('scaler', StandardScaler()), ('model', SVR(kernel = 'linear'))])
pipe_svr.fit(X_train, y_train)
y_svr = pipe_svr.predict(X_test)
r2_svr= pipe_svr.score(X_test, y_test)
rmse_svr = -root_mean_squared_error(y_test, y_svr)
mae_svr = -mean_absolute_error(y_test, y_svr)
print(f'SVR negative R-Squared with linear kernel = {r2_svr:.3f}')
print(f'SVR negative RMSE with linear kernel = {rmse_svr:.3f}')
print(f'SVR negative MAE with linear kernel = {mae_svr:.3f}')

SVR negative R-Squared with linear kernel = 0.932
SVR negative RMSE with linear kernel = -4.466
SVR negative MAE with linear kernel = -3.549


In [ ]:
# __ fitting the data with a support vector regressive model and its r2 parameter

from sklearn.svm import SVR

# creation of the pipeline object with standard scaler normalization, and SVR regressive estimator with a radial basis function kernel
pipe_svr = Pipeline(steps = [('scaler', StandardScaler()), ('model', SVR(kernel = 'rbf'))])
pipe_svr.fit(X_train, y_train)
y_svr = pipe_svr.predict(X_test)
r2_svr= pipe_svr.score(X_test, y_test)
rmse_svr = -root_mean_squared_error(y_test, y_svr)
mae_svr = -mean_absolute_error(y_test, y_svr)
print(f'SVR R-Squared with rbf kernel = {r2_svr:.3f}')
print(f'SVR negative RMSE with rbf kernel = {rmse_svr:.3f}')
print(f'SVR negative MAE with rbf kernel = {mae_svr:.3f}')

SVR R-Squared with rbf kernel = 0.943
SVR negative RMSE with rbf kernel = -4.084
SVR negative MAE with rbf kernel = -3.137


In [29]:
# Gradient Boosting Regression model
pip_gb = Pipeline(steps = [('scaler', StandardScaler()), ('model', GradientBoostingRegressor(n_estimators = 100))])
pip_gb.fit(X_train, y_train)
y_gb = pip_gb.predict(X_test)
r2_gb = r2_score(y_test, y_gb)
rmse_gb = -root_mean_squared_error(y_test, y_gb)
mae_gb = -mean_absolute_error(y_test, y_gb)

print(f'SVR R-Squared with rbf kernel = {r2_gb:.3f}')
print(f'SVR negative RMSE with rbf kernel = {rmse_gb:.3f}')
print(f'SVR negative MAE with rbf kernel = {mae_gb:.3f}')

SVR R-Squared with rbf kernel = 0.952
SVR negative RMSE with rbf kernel = -3.729
SVR negative MAE with rbf kernel = -2.890


In [ ]:
# __ fitting the data with a Random Forest regressive model and its r2 parameter

from sklearn.ensemble import RandomForestRegressor

# creation of the pipeline object with standard scaler normalization, and a random forest estimator with a 100 number of trees
pipe_rf = Pipeline(steps = [('scaler', StandardScaler()), ('model', RandomForestRegressor (n_estimators = 100))])
pipe_rf.fit(X_train, y_train)
y_rf = pipe_rf.predict(X_test)
r2_rf= pipe_rf.score(X_test, y_test)
rmse_rf = -root_mean_squared_error(y_test, y_rf)
mae_rf = -mean_absolute_error(y_test, y_rf)
print(f'random forest R-Squared = {r2_rf:.3f}')
print(f'random forest negative RMSE = {rmse_rf:.3f}')
print(f'random forest negative MAE = {mae_rf:.3f}')

random forest R-Squared = 0.964
random forest negative RMSE = -3.258
random forest negative MAE = -2.398


Considering R2, negative RMSE and negative MAE of all the models, the random forest seems to be the best model for this dataset, because it has the highest
r2, negative RMSE and negative MAE parameters among the other models. the 0.964 value of r-squared quantity shows that 0.964 percent of the energy output (PE)
is defined by this model, and needless to say that the remaining ~0.03 could not be modelled.

The difference between RMSE AND MAE implies the presence of outliers in the dataset. The higher the difference, the higher the number of outliers.